db/db_impl.cc

```diff
@@ -4759,18 +4759,18 @@ Status DBImpl::WriteImpl(const WriteOptions& write_options,
       if (!cfd->mem()->IsEmpty()) {
         // We only consider active mem table, hoping immutable memtable is
         // already in the process of flushing.
-        size_t cfd_size = cfd->mem()->ApproximateMemoryUsage();
-        if (largest_cfd == nullptr || cfd_size > largest_cfd_size) {
-          largest_cfd = cfd;
-          largest_cfd_size = cfd_size;
+        uint64_t seq = cfd->mem()->GetLatestSeqWhenCreate();
```

Comment:
If we want to make sure we always effectively enforce we try the flush the memtable to free the log files, GetLatestSeqWhenCreate() is safer, because it is closer to how we determine whether a log file can be dropped. We only track the log files to be kept while we switch the memtable and never tracked which log file the first key of the memtable goes to. If a new memtable for CF1 is crated while with log file 00003.log, but the first key to CF1 is not inserted until 00004.log, 000003 is still being held util CF1 is flushed. If you want us to guarantee 000003.log is flushed as soon as possible, a switch sequence ID is a safer way. If you are interested in reading the code, here is where you can start: https://github.com/facebook/rocksdb/blob/5.3.fb/db/db_impl.cc#L766 It's pretty compliated though.

Also, GetFirstSequenceNumber() is a complicated variable. It is updated using a CAS logic and I searched the code and no code is relying on the value of it (except 0 or 1) in critical logic. I'm wary of depending on a complicated logic.

I hope we can keep the implementation simply so it is less prone to unexpected behavior. Would this approach generally satisfy your requirement?